### ISSUE TITLE
Error on startup after 12.1 upgrade

### ISSUE THREAD

**Issue by richarth** [umbraco/Umbraco-CMS#14658]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

12.1

### Bug summary

We had a v11 site which we upgraded successfully to v12.  We've just tried to upgrade the site to v12.1 and are encountering the following error when trying to start the site:

`System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Umbraco.Cms.Persistence.EFCore.Migrations.IMigrationProvider Lifetime: Singleton ImplementationType: Umbraco.Cms.Persistence.EFCore.SqlServer.SqlServerMigrationProvider': Unable to resolve service for type 'Microsoft.EntityFrameworkCore.IDbContextFactory`1[Umbraco.Cms.Persistence.EFCore.UmbracoDbContext]' while attempting to activate 'Umbraco.Cms.Persistence.EFCore.SqlServer.SqlServerMigrationProvider'.) (Error while validating the service descriptor 'ServiceType: Umbraco.Cms.Persistence.EFCore.Migrations.IMigrationProvider Lifetime: Singleton ImplementationType: Umbraco.Cms.Persistence.EFCore.Sqlite.SqliteMigrationProvider': Unable to resolve service for type 'Microsoft.EntityFrameworkCore.IDbContextFactory`1[Umbraco.Cms.Persistence.EFCore.UmbracoDbContext]' while attempting to activate 'Umbraco.Cms.Persistence.EFCore.Sqlite.SqliteMigrationProvider'.) (Error while validating the service descriptor 'ServiceType: Umbraco.Cms.Infrastructure.Migrations.IEFCoreMigrationExecutor Lifetime: Singleton ImplementationType: Umbraco.Cms.Persistence.EFCore.EfCoreMigrationExecutor': Unable to resolve service for type 'Microsoft.EntityFrameworkCore.IDbContextFactory`1[Umbraco.Cms.Persistence.EFCore.UmbracoDbContext]' while attempting to activate 'Umbraco.Cms.Persistence.EFCore.SqlServer.SqlServerMigrationProvider'.) (Error while validating the service descriptor 'ServiceType: Umbraco.Cms.Core.Events.INotificationAsyncHandler`1[Umbraco.Cms.Infrastructure.Migrations.Notifications.DatabaseSchemaAndDataCreatedNotification] Lifetime: Transient ImplementationType: Umbraco.Cms.Persistence.EFCore.Composition.EFCoreCreateTablesNotificationHandler': Unable to resolve service for type 'Microsoft.EntityFrameworkCore.IDbContextFactory`1[Umbraco.Cms.Persistence.EFCore.UmbracoDbContext]' while attempting to activate 'Umbraco.Cms.Persistence.EFCore.SqlServer.SqlServerMigrationProvider'.) (Error while validating the service descriptor 'ServiceType: Umbraco.Cms.Core.Events.INotificationAsyncHandler`1[Umbraco.Cms.Core.Notifications.UnattendedInstallNotification] Lifetime: Transient ImplementationType: Umbraco.Cms.Persistence.EFCore.Composition.EFCoreCreateTablesNotificationHandler': Unable to resolve service for type 'Microsoft.EntityFrameworkCore.IDbContextFactory`1[Umbraco.Cms.Persistence.EFCore.UmbracoDbContext]' while attempting to activate 'Umbraco.Cms.Persistence.EFCore.SqlServer.SqlServerMigrationProvider'.)'`

I don't know if anybody has any ideas?

### Steps to reproduce

Upgrade a v11 site to v12, upgrade to v12.1 in Visual Studio then attempt to run the site.

### Expected result / actual result

The site should run and the upgrade process should run.

---
_This item has been added to our backlog [AB#31902](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/31902)_

**Comment by arknu** [comment]
We're seeing this as well upgrading from 12.0 to 12.1

**Comment by rammi987** [comment]
See the same as well upgrading from 12.0.1 to 12.1.0

**Comment by rammi987** [comment]
Hello @richarth and @arknu 
I solved the issue for my project going from 12.0.1 to 12.1.0.

Can you guys inside your project find the Startup.cs file and inside that file try to search for the pisce of code `.AddDeliveryApi()`
If you missing this line inside the Startup.cs file then search for `.AddWebsite()` and the add the `.AddDeliveryApi()` as in the next of the pipe chain so you end up with this in the method `ConfigureServices`

```
var umbraco = services.AddUmbraco(_env, _config)
                .AddBackOffice()
                .AddWebsite()
                .AddDeliveryApi()
                .AddComposers()
                .Build();
```

and the try to rerun the website. It fix my error.

If you have the .AddDeliveryApi() inside your startup file, please also return then i can debug futher

**Comment by kjac** [comment]
Hi everyone. That's really interesting 🤔 

I can reproduce this on a fresh 12.1 install by removing `.AddDeliveryApi()` from `ConfigureServices` in `Startup`. That's ... unexpected to say the least. 

We'll look into this 🔍

**Comment by kjac** [comment]
To clarify for anyone just tuning in. `ConfigureServices` in `Startup` must contain `.AddDeliveryApi()` - like so:

```cs
public void ConfigureServices(IServiceCollection services)
{
	services.AddUmbraco(_env, _config)
		.AddBackOffice()
		.AddWebsite()
		.AddDeliveryApi()
		.AddComposers()
		.Build();
}
```

Adding this line does NOT turn on the Delivery API on its own, so it will not cause unexpected data leaks. The Delivery API must be explicitly enabled in `appsettings.json`. See [the docs](https://docs.umbraco.com/umbraco-cms/reference/content-delivery-api) for more info.

**Comment by kjac** [comment]
...thanks for the quick investigation efforts, @rammi987 👍

**Comment by richarth** [comment]
Hi,

Thanks @rammi987.  I can confirm the site was missing `.AddDeliveryApi()`.  Adding that has gotten the site running again.

**Comment by rammi987** [comment]
@kjac found some
[inside the Umbraco.Cms.Api.Delivery.DependencyInjection.UmbracoBuilderExtensions](https://github.com/umbraco/Umbraco-CMS/blob/0edee82b0b92876ef336a04fa82eb48832302523/src/Umbraco.Cms.Api.Delivery/DependencyInjection/UmbracoBuilderExtensions.cs#L33) add in the `UmbracoDbContext` Context.

So deleting the line `AddDeliveryApi()` will acully do the the EF Core context of type UmbracoDbContext never exiting in the DI.
So when Umbraco is starting is hit the composer https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Cms.Persistence.EFCore/Composition/UmbracoEFCoreComposer.cs

that dependens on the `UmbracoDbContext` but again never added to the ServiceProvider scope.

**Comment by rammi987** [comment]
Even after Upgrade NEVER delete the AddDeliveryApi() line, the exception will return 💀

**Comment by thomas-jf** [comment]
Yes, adding `.AddDeliveryApi()` in Startup fixes the issue (updating from 12.0.1 to 12.1.0)

**Comment by JasonElkin** [comment]
I've seen @Migaroez PR #14667 so Startup.cs may not be the best place for `AddDeliveryApi()` anyway, but...

It would be a good idea to have some kind of check/safeguard in place to sound the klaxons when `Startup.cs` (or `Program.cs`) gets modified as it will always throw a spanner in the works for upgrades - changing it should be avoided unless absolutely necessary.

If it _is_ necessary then I'd expect to see that reflected in the release notes, and in the [version specific upgrade docs](https://docs.umbraco.com/umbraco-cms/fundamentals/setup/upgrading/version-specific) and for everyone that knows about it to make a lot of noise about it before the release.

**Comment by nikcio** [comment]
To clarify this issue stems from the https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Cms.Persistence.EFCore/Composition/UmbracoEFCoreComposer.cs

Because here some services are added that rely on the DBContext but for some reason, it was decided that the DBContext should be added to the services explicitly in the Delivery API project 🤔 

@kjac This was also one of the things I mentioned here: https://github.com/umbraco/Umbraco-CMS/pull/14449#issuecomment-1637044620 😅 

And in this PR: https://github.com/umbraco/Umbraco-CMS/pull/14563

**Comment by Ambertvu** [comment]
Just a minor addition, I encountered this issue as well, but only when running in Production mode, locally it would run and upgrade fine.

**Comment by nikolajlauridsen** [comment]
Heya, this has been fixed in #14678 which is @nikcio's PR but without the breaking change 😄 So big thank you for that 😄 

This change has already been released in 12.1.1, which is available on Nuget, so I'll go ahead and close this issue, thanks again.
